keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,468 @@
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The performance of kube-proxy's iptables mode in large clusters could
+be greatly increased by optimizing its calls to `iptables-restore`.
+Although this is a small change in terms of code size, it has a large
+effect on kube-proxy operation, and could potentially completely break
+clusters (if the code was buggy), so we have decided to treat it as a
+KEP-able feature.
+
+## Motivation
+
+### Goals
+
+- Improve the performance of kube-proxy in iptables mode, by optimizing
+  its calls to `iptables-restore` by not including rules that haven't
+  changed.
+
+### Non-Goals
+
+- Everything else.
+
+## Proposal
+
+### General Plan
+
+`iptables-restore` has the constraint that if you want to specify
+_any_ rule in a given iptables chain, then you must specify _every_
+rule in that chain. However, if you want to leave a chain completely
+unchanged, then you can just fail to mention any rules in that chain,
+and it will be left untouched. (If you want to delete a chain you have
```

Comment:
are deletes considered here? it seems to be only about updates

Reply:
The KEP doesn't explicitly discuss deletes anywhere else, but at the level of abstraction in the discussion here, they can just be thought of as a different kind of update.

The code PR handles deletes correctly (and the unit test tests that).